server/src/main/java/com/example/tyfserver/oauth2/service/OAuth2Service.java

```diff
@@ -0,0 +1,123 @@
+package com.example.tyfserver.oauth2.service;
+
+import com.example.tyfserver.member.repository.MemberRepository;
+import com.example.tyfserver.oauth2.domain.Oauth2Type;
+import lombok.RequiredArgsConstructor;
+import org.json.JSONObject;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.stereotype.Service;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+@RequiredArgsConstructor
+public class OAuth2Service {
+    private final MemberRepository memberRepository;
+
+    private final Oauth2Type kakao;
+    private final Oauth2Type google;
+    private final Oauth2Type naver;
+
+    public String loginKakao(String code) {
+        return login(code, kakao);
+    }
+
+    public String loginGoogle(String code) {
+        return login(code, google);
+    }
+
+    public String loginNaver(String code) {
+        return login(code, naver);
+    }
+
+    public String signUpKakao(String code) {
+        return signUp(code, kakao);
+    }
+
+    public String signUpGoogle(String code) {
+        return signUp(code, google);
+    }
+
+    public String signUpNaver(String code) {
+        return signUp(code, naver);
+    }
+
+    private String login(String code, Oauth2Type oAuth2Type) {
+        final String accessToken = requestAccessToken(code, oAuth2Type);
+        final String email = requestEmail(accessToken, oAuth2Type);
+
+        memberRepository.findByEmail(email)
+                .orElseThrow(() -> new RuntimeException("가입되어 있지 않은 유저")); // todo 예외클래스, 메시지
+
+        // todo JWT 발급
+        return email;
+    }
+
+    private String signUp(String code, Oauth2Type oAuth2Type) {
+        final String accessToken = requestAccessToken(code, oAuth2Type);
+        final String email = requestEmail(accessToken, oAuth2Type);
+
+        // 멤버가 있는지 확인
+        // 1. 멤버가 없음: 가입
+        // 2. 멤버가 있고 다른 타입: 어떤 타입으로 가입되어있는지 안내
+        // 3. 멤버가 있고 같은 타입: 이미 가입되어있다 안내하고 로그인
+
+        // todo JWT 발급
+        return email;
+    }
+
+    private String requestAccessToken(String code, Oauth2Type oAuth2Type) {
+        final RestTemplate restTemplate = new RestTemplate();
+
+        final String body = restTemplate.postForObject(
+                oAuth2Type.getAccessTokenApi(),
+                generateAccessTokenRequest(code, oAuth2Type),
+                String.class
+        );
+
+        final JSONObject jsonObject = new JSONObject(body);
+        return jsonObject.getString("access_token");
+    }
+
+    private HttpEntity<MultiValueMap<String, String>> generateAccessTokenRequest(String code, Oauth2Type oAuth2Type) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Content-type", "application/x-www-form-urlencoded;charset=utf-8");
```

Comment:
```suggestion
        headers.add(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded;charset=utf-8");
```
이미 제공하는`상수`로 대체하는 것이 더 좋을 것 같아 👍

Reply:
```suggestion
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
        headers.setContentType(MediaType.APPLICATION_JSON_UTF8);
```
이렇게 대체할 수도 있을 것 같어!   (APPLICATION_JSON_UTF8는 이제 안쓴다고 하는데 크롬과 같은 대부분의 브라우저가 잘 지원해서 불필요하다는거 같은데, 이거 맞나 다들 확인좀 ㅋㅋㅋㅋㅋ 영어 앞에서는 한 없이 작아지네 🥲)

Reply:
어디건지 기억안나는데 api 문서 예시에 저렇게 써있어서 그냥 저렇게 한 것 같음! 나도 그랬어가지고 ㅋㅋ

Reply:
아하아하